docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,206 @@
+If the URL returns something other than an archive, it is downloaded to `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}`.
+
+Both `DAEMON_HOME` and `DAEMON_NAME` are [environment variables used to configure Cosmovisor](https://github.com/cosmos/cosmos-sdk/blob/cosmovisor/v1.0.0/cosmovisor/README.md#command-line-arguments-and-environment-variables).
+
+Currently, there is no mechanism that makes Cosmovisor run a command after the upgraded chain has been restarted.
+
+## Decision
+
+### Protobuf Updates
+
+We will define a new message for providing upgrade instructions and add it as a new field in the `Plan` message.
+The upgrade instructions will contain a list of artifacts available for each platform.
+It will also allow for the definition of a pre-run and post-run command.
+
+```protobuf
+message Plan {
+  // ... (existing fields)
+
+  UpgradeInstructions upgrade = 6;
+}
+```
+
+The new `UpgradeInstructions upgrade` field MUST be optional.
+
+```protobuf
+message UpgradeInstructions {
+  string pre_run              = 1;
+  string post_run             = 2;
+  repeated Artifact artifacts = 3;
+  string description          = 4;
+}
+```
+
+All fields in the `UpgradeInstructions` are optional.
+- `pre_run` is a command to run prior to the upgraded chain restarting.
+  If defined, then the app supervisors (e.g. Cosmovisor) MUST NOT run `app pre-run`.
+  If defined, this command MUST behave the same as the current [pre-upgrade](https://github.com/cosmos/cosmos-sdk/blob/v0.44.5/docs/migrations/pre-upgrade.md) command.
+  The working directory this command runs from MUST be `{DAEMON_HOME}/cosmovisor/{upgrade name}`.
+- `post_run` is a command to run after the upgraded chain has been started. If defined, this command MUST be only executed once.
+  The output and exit code SHOULD be logged but SHOULD NOT affect the running of the upgraded chain.
+  The working directory this command runs from MUST be `{DAEMON_HOME}/cosmovisor/{upgrade name}`.
+- `artifacts` define items to be downloaded.
+  It SHOULD have only one entry per platform.
+- `description` contains human-readable information about the upgrade and might contain references to external resources.
+  It SHOULD NOT be used for structured processing information.
+
+```protobuf
+message Artifact {
+  string platform      = 1;
+  string url           = 2;
+  string checksum      = 3;
+  string checksum_algo = 4;
+}
+```
+
+- `platform` is a required string that SHOULD be in the format `{OS}/{CPU}`, e.g. `"linux/amd64"`.
+  The string `"any"` SHOULD also be allowed.
+  An `Artifact` with a `platform` of `"any"` SHOULD be used as a fallback when a specific `{OS}/{CPU}` entry is not found.
+  That is, if an `Artifact` exists with a `platform` that matches the system's OS and CPU, that should be used;
+  otherwise, if an `Artifact` exists with a `platform` of `any`, that should be used;
+  otherwise no artifact should be downloaded.
```

Comment:
Maybe we should return an error if there are artifacts, but none matches the running platform?

Reply:
So, if there is a common, shared artifact, then it has to be copied to all platforms? Let's make a note about that.

Reply:
This description is what Cosmovisor currently does. If this new stuff is going to have different behavior, we should use a word other than `any` for it.

I think current behavior is sufficient, though. We can maybe update Cosmovisor to issue a warning if there are multiple platforms, but the desired one isn't found and "any" is used instead, but I don't think that behavior needs to be part of spec.